packages/aws-cdk/toolkit/lib/cli-io-host.ts

```diff
@@ -0,0 +1,101 @@
+import * as chalk from 'chalk';
+
+/**
+ * Basic message structure for toolkit notifications.
+ * Messages are emitted by the toolkit and handled by the IoHost.
+ */
+interface IoMessage {
+  time: Date;
+  level: IoMessageLevel;
+  action: IoAction;
+  code: string;
+  message: string;
+  forceStdout?: boolean;
+  // Optionally specify Chalk style for stdout/stderr, if TTY is enabled
+  style?: ((str: string) => string);
+}
+
+export type IoMessageLevel = 'error' | 'warn' | 'info' | 'debug' | 'trace';
+
+export type IoAction = 'synth' | 'list' | 'deploy' | 'destroy';
+
+interface CliIoHostOptions {
+  useTTY?: boolean;
+  ci?: boolean;
+}
+
+/**
+ * A simple IO host for the CLI that writes messages to the console.
+ */
+export class CliIoHost {
+  private readonly pretty_messages: boolean;
+  private readonly ci: boolean;
+
+  constructor(options: CliIoHostOptions) {
+    this.pretty_messages = options.useTTY ?? process.stdout.isTTY ?? false;
+    this.ci = options.ci ?? false;
+  }
+
+  /**
+   * Notifies the host of a message.
+   * The caller waits until the notification completes.
+   */
+  async notify(msg: IoMessage): Promise<void> {
+    const output = this.formatMessage(msg);
+
+    const stream = this.getStream(msg.level, msg.forceStdout ?? false);
+
+    return new Promise((resolve, reject) => {
+      stream.write(output + '\n', (err) => {
+        if (err) {
+          reject(err);
+        } else {
+          resolve();
+        }
+      });
+    });
+  }
+
+  /**
+   * Determines which output stream to use based on log level and configuration.
+   */
+  private getStream(level: IoMessageLevel, forceStdout: boolean) {
+    if (forceStdout) {
+      return process.stdout;
+    }
+    if (level == 'error') return process.stderr;
+    return this.ci ? process.stdout : process.stderr;
```

Comment:
can you document the reasoning for why we are sending certain things to stderr and certain things to stdout?

Reply:
The honest answer is "I'm preserving the way it worked in the past, but I'll add a history lesson here because it's truly not obvious why this is here or why we default to stderr

Reply:
So I dug through the annals of CDK history until I hit [the very beginning](https://github.com/aws/aws-cdk/pull/1/files#diff-0b1ef5d561ead34a93ad9b088d8f5a4f99083654c72381b5d4bb34682c8a00c9)

I'm gonna give myself a backlog item to consolidate our ci check to the CliIoHost and remove that logic from the CLI. It's not relevant to this PR though so I'm not gonna do that here,

Reply:
its ok that this is because of legacy reasons. i want documentation here to explain the following:

forceStdout -> stdout regardless of log level
ci -> stdout except for log level error
non-ci -> stderr